actix-session/src/storage/interface.rs

```diff
@@ -0,0 +1,67 @@
+use super::SessionKey;
+use std::collections::HashMap;
+use time::Duration;
+
+pub(crate) type SessionState = HashMap<String, String>;
+
+/// The interface to retrieve and save the current session data from/to the
+/// chosen storage backend.
+///
+/// You can provide your own custom session store backend by implementing this trait.
+#[async_trait::async_trait(?Send)]
+pub trait SessionStore {
+    /// Load the session state associated to a session key.
+    async fn load(&self, session_key: &SessionKey) -> Result<Option<SessionState>, LoadError>;
+
+    /// Persist the session state for a newly created session.
+    /// It returns the corresponding session key.
+    async fn save(
+        &self,
+        session_state: SessionState,
+        ttl: &Duration,
+    ) -> Result<SessionKey, SaveError>;
+
+    /// Update the session state associated to a pre-existing session key.
+    async fn update(
+        &self,
+        session_key: SessionKey,
+        session_state: SessionState,
+        ttl: &Duration,
+    ) -> Result<SessionKey, UpdateError>;
+
+    /// Delete a session from the store.
+    async fn delete(&self, session_key: &SessionKey) -> Result<(), anyhow::Error>;
+}
+
+#[derive(thiserror::Error, Debug)]
+/// Possible failures modes for [`SessionStore::load`].
+pub enum LoadError {
```

Comment:
Having a series of Error types related to sessions, rather than a single Error type, is over-engineered for a library.  What is the rationale?

Reply:
Note that there was a manual implementation of the Error trait in this library that you replaced by adding `anyhow` and `thiserror` as dependencies.  Error proc macros have a tradeoff of convenience for compilation time.  Let's not add proc macros nor new dependencies where they aren't needed, ok?

Reply:
> Having a series of Error types related to sessions, rather than a single Error type, is over-engineered for this particular library. Considering there was only an InsertError to begin with, it's understand why you continued with that theme, but we will be fine with consolidating to a single enum going forward. Curious what your thoughts are, though.

I think that having a single error enum for distinct operations with different failure modes is anti-pattern, it leads to confusion on the caller side.

Let's go with a specific example: `delete` can never fail due to a serialization error, because there is no serialization in the logic it has to execute.
If we were to have a single `SessionStoreError` with a `GenericError` and `SerializationError` variant we would have to handle the latter as well. This either leads to unnecessary logic on the caller side or, equally problematic, the caller makes an assumption on the implementation details of the `SessionStore` by slapping an `unreachable!` macro on the `SerializationError` branch. The implementation changes at a later date, it now involves serialization, there is no compilation error and the caller starts seeing panics at runtime.

In the specific case of `SessionStore` it is worth stressing that these operations and errors are not going to be a concern for 99% of application developers using this library - the only people who are going to interact with the `SessionStore` trait are store implementers and us, in the middleware. The more precise the interface, the easier it is for a store implementer to provide a store that behaves predictably because its failures are handled correctly by our middleware.

Reply:
> Note that there was a manual implementation of the Error trait in this library that you replaced by adding anyhow and thiserror as dependencies. Error proc macros have a tradeoff of convenience for compilation time. Let's not add proc macros nor new dependencies where they aren't needed, ok?

I have added `anyhow` to allow store implementers to upcast whatever error is returned by the specific drivers they are using (e.g. `actix-redis` with `redis-async` or `redis`) into a common representation which we can then handle in our middleware code. `anyhow` is the best solution at the moment within the ecosystem, especially if we consider the need to preserve the error cause chain.
I'd be happy to consider alternatives, but I think the addition pulls its own weight in terms of ergonomics for store implementers. `anyhow` does not bring any transitive dependency, it is quite lightweight - I can run some compilation benchmarks if needed.

Re: `thiserror` - we could replace indeed it with manual implementations of `Display` and `std::error::Error`.
These are the dependencies brought in by `thiserror`:
```
── thiserror v1.0.30
│   └── thiserror-impl v1.0.30 (proc-macro)
│       ├── proc-macro2 v1.0.34 (*)
│       ├── quote v1.0.10 (*)
│       └── syn v1.0.82 (*)
```
These three crates are already in our dependency tree, even when no optional features are enabled. They are used by `serde` (derive), by `actix-web` (via derive-more, cookie, etc.), by `actix-server`, etc. - you can see it by launching `cargo tree -i syn` locally. 
So the only added "weight" in terms of dependency is the code in `thiserror` itself. Happy to remove it if you feel strongly about it, but the argument for bloat has nothing to do with proc-macros here.

Reply:
RE: error libs:
- We've gotten by with `Box<dyn StdError>` in -web, would this not do here also vs adding `anyhow`?
- I would prefer using `derive_more` macros for Error impls.

Reply:
> I would prefer using derive_more macros for Error impls.

I swapped `thiserror` for `derive_more` in 4a5fe0ee6a4e15e11571243d6842c7f67f773892

Reply:
> We've gotten by with Box<dyn StdError> in -web, would this not do here also vs adding anyhow?

There are two main pros in favour of using `anyhow::Error` instead of `Box<dyn std::error::Error>`:
- backtraces are always available (if enabled), even on stable;
- the `Debug` implementation traverses the entire error source chain, providing the necessary context to interpret what went wrong. 

The main con is that it requires the underlying error type to be `Send + Sync`.